File: traditional-api/src/controllers/account.controller.js
const accountModel = require('../models/account');
const { getTransactionHistory } = require('../engines/ledger');

/**
 * POST /api/v1/accounts
 *
 * Create an additional account for the logged in customer.
 * They already have a savings account from registration.
 * This is for when they want a current account or wallet too.
 */
const createAccount = async (req, res) => {
  try {
    const { accountType, accountName, currency } = req.body;

    // Only these types are allowed
    const validTypes = ['savings', 'current', 'wallet'];
    if (!validTypes.includes(accountType)) {
      return res.status(400).json({
        success: false,
        message: `Invalid account type. Must be one of: ${validTypes.join(', ')}`,
      });
    }

    const account = await accountModel.createAccount({
      customerId:  req.customer.id,
      accountType,
      accountName: accountName ||
        `${req.customer.first_name} ${req.customer.last_name} - ${accountType}`,
      currency:    currency || 'USD',
    });

    return res.status(201).json({
      success: true,
      message: 'Account created successfully.',
      data: {
        account: { ...account, balance: 0 }
      },
    });

  } catch (err) {
    console.error('Create account error:', err);
    return res.status(500).json({
      success: false,
      message: 'Failed to create account.',
    });
  }
};

/**
 * GET /api/v1/accounts
 *
 * Get all accounts for the logged in customer.
 * Each account includes its live balance.
 * This is the home screen of the banking app.
 */
const getMyAccounts = async (req, res) => {
  try {
    const accounts = await accountModel.getCustomerAccounts(req.customer.id);

    return res.json({
      success: true,
      data: {
        accounts,
        count: accounts.length,
      },
    });

  } catch (err) {
    console.error('Get accounts error:', err);
    return res.status(500).json({
      success: false,
      message: 'Failed to fetch accounts.',
    });
  }
};

/**
 * GET /api/v1/accounts/:id
 *
 * Get a single account with its live balance.
 * Security: only returns the account if it belongs
 * to the logged in customer.
 */
const getAccount = async (req, res) => {
  try {
    // Pass both accountId AND customerId
    // Model will only return account if both match
    const account = await accountModel.getAccountById(
      req.params.id,
      req.customer.id
    );

    if (!account) {
      return res.status(404).json({
        success: false,
        message: 'Account not found.',
      });
    }

    return res.json({
      success: true,
      data: { account },
    });

  } catch (err) {
    console.error('Get account error:', err);
    return res.status(500).json({
      success: false,
      message: 'Failed to fetch account.',
    });
  }
};

/**
 * GET /api/v1/accounts/:id/balance
 *
 * Get just the balance for quick checks.
 * Useful for the iOS app to refresh balance
 * without loading all account details.
 */
const getBalance = async (req, res) => {
  try {
    const account = await accountModel.getAccountById(
      req.params.id,
      req.customer.id
    );

    if (!account) {
      return res.status(404).json({
        success: false,
        message: 'Account not found.',
      });
    }

    return res.json({
      success: true,
      data: {
        accountId:     account.id,
        accountNumber: account.account_number,
        balance:       parseFloat(account.balance),
        currency:      account.currency,
      },
    });

  } catch (err) {
    console.error('Get balance error:', err);
    return res.status(500).json({
      success: false,
      message: 'Failed to fetch balance.',
    });
  }
};

/**
 * GET /api/v1/accounts/:id/transactions
 *
 * Transaction history for an account.
 * This is what the app shows when you tap on an account.
 * Supports pagination via limit and offset query params.
 *
 * Example: /accounts/:id/transactions?limit=20&offset=0
 */
const getAccountTransactions = async (req, res) => {
  try {
    // First verify this account belongs to the customer
    const account = await accountModel.getAccountById(
      req.params.id,
      req.customer.id
    );

    if (!account) {
      return res.status(404).json({
        success: false,
        message: 'Account not found.',
      });
    }

    // Parse pagination params from URL
    // Default: 20 transactions per page starting from 0
    const limit  = parseInt(req.query.limit)  || 20;
    const offset = parseInt(req.query.offset) || 0;

    const transactions = await getTransactionHistory(
      req.params.id,
      { limit, offset }
    );

    return res.json({
      success: true,
      data: {
        transactions,
        limit,
        offset,
      },
    });

  } catch (err) {
    console.error('Get transactions error:', err);
    return res.status(500).json({
      success: false,
      message: 'Failed to fetch transactions.',
    });
  }
};

/**
 * PATCH /api/v1/accounts/:id/status
 *
 * Freeze, unfreeze or mark an account dormant.
 * Customer can freeze their own account
 * (e.g. if they lose their card).
 */
const updateStatus = async (req, res) => {
  try {
    const { status } = req.body;

    const validStatuses = ['active', 'frozen', 'dormant'];
    if (!validStatuses.includes(status)) {
      return res.status(400).json({
        success: false,
        message: `Invalid status. Must be one of: ${validStatuses.join(', ')}`,
      });
    }

    // Verify account belongs to customer before changing anything
    const account = await accountModel.getAccountById(
      req.params.id,
      req.customer.id
    );

    if (!account) {
      return res.status(404).json({
        success: false,
        message: 'Account not found.',
      });
    }

    const updated = await accountModel.updateAccountStatus(
      req.params.id,
      status,
      req.customer.id
    );

    return res.json({
      success: true,
      message: `Account ${status} successfully.`,
      data: { account: updated },
    });

  } catch (err) {
    console.error('Update status error:', err);
    return res.status(500).json({
      success: false,
      message: 'Failed to update account status.',
    });
  }
};

module.exports = {
  createAccount,
  getMyAccounts,
  getAccount,
  getBalance,
  getAccountTransactions,
  updateStatus,
};

File: banking-gateway/src/main/java/com/corebanking/gateway/controller/AccountController.java
package com.corebanking.gateway.controller;

import com.corebanking.gateway.model.response.ApiResponse;
import com.corebanking.gateway.service.AccountService;
import com.corebanking.gateway.service.TokenStoreService;
import jakarta.servlet.http.HttpServletRequest;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.Map;

/**
 * ACCOUNT CONTROLLER
 *
 * Handles all account-related endpoints.
 * Extracts the CBS token from TokenStore
 * and forwards requests to AccountService.
 *
 * GET  /api/v1/accounts                     → all accounts
 * GET  /api/v1/accounts/:id                 → single account
 * GET  /api/v1/accounts/:id/balance         → balance only
 * GET  /api/v1/accounts/:id/transactions    → history
 * POST /api/v1/accounts                     → create account
 * PATCH /api/v1/accounts/:id/status         → freeze/unfreeze
 */
@Slf4j
@RestController
@RequestMapping("/api/v1/accounts")
@RequiredArgsConstructor
public class AccountController {

    private final AccountService accountService;
    private final TokenStoreService tokenStoreService;

    /**
     * GET /api/v1/accounts
     * Home screen — all accounts with live balances
     */
    @GetMapping
    public ResponseEntity<ApiResponse<Object>> getMyAccounts(
            @RequestAttribute("customerId") String customerId
    ) {
        try {
            String cbsToken = tokenStoreService.getCbsToken(customerId);
            Map<String, Object> result = accountService.getMyAccounts(cbsToken);
            return ResponseEntity.ok(
                    ApiResponse.success("Accounts fetched.", result.get("data"))
            );
        } catch (Exception e) {
            log.error("Get accounts error: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error(e.getMessage()));
        }
    }

    /**
     * GET /api/v1/accounts/:id
     */
    @GetMapping("/{accountId}")
    public ResponseEntity<ApiResponse<Object>> getAccount(
            @PathVariable String accountId,
            @RequestAttribute("customerId") String customerId
    ) {
        try {
            String cbsToken = tokenStoreService.getCbsToken(customerId);
            Map<String, Object> result = accountService.getAccount(
                    accountId, cbsToken
            );
            return ResponseEntity.ok(
                    ApiResponse.success("Account fetched.", result.get("data"))
            );
        } catch (Exception e) {
            log.error("Get account error: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error(e.getMessage()));
        }
    }

    /**
     * GET /api/v1/accounts/:id/balance
     */
    @GetMapping("/{accountId}/balance")
    public ResponseEntity<ApiResponse<Object>> getBalance(
            @PathVariable String accountId,
            @RequestAttribute("customerId") String customerId
    ) {
        try {
            String cbsToken = tokenStoreService.getCbsToken(customerId);
            Map<String, Object> result = accountService.getBalance(
                    accountId, cbsToken
            );
            return ResponseEntity.ok(
                    ApiResponse.success("Balance fetched.", result.get("data"))
            );
        } catch (Exception e) {
            log.error("Get balance error: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error(e.getMessage()));
        }
    }

    /**
     * GET /api/v1/accounts/:id/transactions
     */
    @GetMapping("/{accountId}/transactions")
    public ResponseEntity<ApiResponse<Object>> getTransactions(
            @PathVariable String accountId,
            @RequestParam(defaultValue = "20") int limit,
            @RequestParam(defaultValue = "0")  int offset,
            @RequestAttribute("customerId") String customerId
    ) {
        try {
            String cbsToken = tokenStoreService.getCbsToken(customerId);
            Map<String, Object> result = accountService.getAccountTransactions(
                    accountId, limit, offset, cbsToken
            );
            return ResponseEntity.ok(
                    ApiResponse.success("Transactions fetched.", result.get("data"))
            );
        } catch (Exception e) {
            log.error("Get transactions error: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error(e.getMessage()));
        }
    }

    /**
     * POST /api/v1/accounts
     * Create additional account
     */
    @PostMapping
    public ResponseEntity<ApiResponse<Object>> createAccount(
            @RequestBody Map<String, Object> requestBody,
            @RequestAttribute("customerId") String customerId
    ) {
        try {
            String cbsToken = tokenStoreService.getCbsToken(customerId);
            Map<String, Object> result = accountService.createAccount(
                    requestBody, cbsToken
            );
            return ResponseEntity.status(201).body(
                    ApiResponse.success("Account created.", result.get("data"))
            );
        } catch (Exception e) {
            log.error("Create account error: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error(e.getMessage()));
        }
    }

    /**
     * PATCH /api/v1/accounts/:id/status
     * Freeze or unfreeze
     */
    @PatchMapping("/{accountId}/status")
    public ResponseEntity<ApiResponse<Object>> updateStatus(
            @PathVariable String accountId,
            @RequestBody Map<String, Object> requestBody,
            @RequestAttribute("customerId") String customerId
    ) {
        try {
            String status   = (String) requestBody.get("status");
            String cbsToken = tokenStoreService.getCbsToken(customerId);
            Map<String, Object> result = accountService.updateAccountStatus(
                    accountId, status, cbsToken
            );
            return ResponseEntity.ok(
                    ApiResponse.success("Account status updated.", result.get("data"))
            );
        } catch (Exception e) {
            log.error("Update status error: {}", e.getMessage());
            return ResponseEntity.badRequest()
                    .body(ApiResponse.error(e.getMessage()));
        }
    }
}